packages/legend-application-query/src/stores/QueryBuilderChangeDetectionState.ts

```diff
@@ -0,0 +1,161 @@
+   * See {@link preComputeGraphElementHashes} for more details
+   */
+  private stateHashCodeKeepAliveComputationDisposers: IDisposer[] = [];
+  private changeDetectionReaction?: IReactionDisposer | undefined;
+
+  constructor(queryBuilderState: QueryBuilderState) {
+    makeObservable(this, {
+      initialQueryHashCode: observable,
+      currentQueryHashCode: observable,
+      stateHashIndex: observable,
+      isBuildingStateHashesIndex: observable,
+      setInitialQueryHashCode: action,
+      setCurrentQueryHashCode: action,
+      detectChanges: flow,
+      setIsBuildingStateHashesIndex: action,
+      buildStateHashesIndex: flow,
+    });
+
+    this.querybuilderState = queryBuilderState;
+  }
+
+  setQueryHashCode(val: string): void {
+    this.queryHashCode = val;
+  }
+
+  setIsEnabled(val: boolean): void {
+    this.isEnabled = val;
+  }
+
+  setInitialQueryHashCode(val: string | undefined): void {
+    this.initialQueryHashCode = val;
+  }
+
+  setCurrentQueryHashCode(val: string | undefined): void {
+    this.currentQueryHashCode = val;
+  }
+
+  setIsBuildingStateHashesIndex(val: boolean): void {
+    this.isBuildingStateHashesIndex = val;
+  }
+
+  async buildHashesIndex(): Promise<string> {
+    return promisify(() => this.querybuilderState.hashCode);
+  }
+
+  *detectChanges(): GeneratorFn<void> {
+    //TODO: Set states to indicate that query has got changed and query results have gone stale
+  }
+
+  *buildStateHashesIndex(): GeneratorFn<void> {
+    this.setIsBuildingStateHashesIndex(true);
+    try {
+      const hashesIndex = (yield this.buildHashesIndex()) as string;
+      this.setInitialQueryHashCode(hashesIndex);
+      this.setIsBuildingStateHashesIndex(false);
+    } catch (error) {
+      assertErrorThrown(error);
+      this.querybuilderState.applicationStore.log.error(
+        LogEvent.create(
+          QUERY_BUILDER_CHANGE_DETECTION_EVENT.CHANGE_DETECTION_FAILURE,
+        ),
+        `Can't build hashes index for query builder`,
+      );
+      this.setInitialQueryHashCode(undefined);
+      this.setIsBuildingStateHashesIndex(false);
+      throw new IllegalStateError(error);
+    } finally {
+      this.setIsBuildingStateHashesIndex(false);
+    }
+  }
+
+  snapshotStateHashesIndex(): string {
+    return this.querybuilderState.hashCode;
+  }
+
+  async preComputeStateHashes(): Promise<void> {
+    const disposers: IDisposer[] = [];
+    await promisify(() => {
+      if (isComputedProp(this.querybuilderState, 'hashCode')) {
+        disposers.push(keepAlive(this.querybuilderState, 'hashCode'));
+        // manually trigger hash code computation
+        this.querybuilderState.hashCode;
+      }
+    });
+    // save the `keepAlive` computation disposers to dispose after we start change detection
+    // so the first call of change detection still get the benefit of computed value
+    this.stateHashCodeKeepAliveComputationDisposers = disposers;
+  }
+
+  start(): void {
```

Comment:
I wonder if you even need this at all anymore. Technically, we don't need this `reaction` block. We are only interested in the very first hash and subsequent hash has been incrementally built up using the tree of `hashCode` you created

Reply:
Let's remove all of these reaction logic. Basically, think of the user of `QueryBuilderChangeDetectionState` all they have to do is:

`await queryBuilderState.changeDetectionState.start()`